### ISSUE TITLE
Syntax highlighting via LPeg

### ISSUE THREAD

**Issue by leavengood** [google/xi-editor#23]
One possible suggestion for the method of syntax highlighting is via [LPeg](http://www.inf.puc-rio.br/%7Eroberto/lpeg/). This is used by the [Vis editor](https://github.com/martanne/vis), which borrowed it from the Scintillua project as described in this [README](https://github.com/martanne/vis/blob/master/lexers/README.md).

You can see all the languages supported in the [Vis lexers directory](https://github.com/martanne/vis/tree/master/lexers).

If you wanted to embed this in xicore there seem to be multiple Rust Lua bindings, such as these two:

https://github.com/kballard/rust-lua
https://github.com/jcmoyer/rust-lua53

Or maybe LPeg could be use from a Lua-based external tool using whatever future RPC mechanism you intend for "plugins". But this may make sense to be done inside the core.

**Comment by raphlinus** [comment]
I am very uninclined to link a scripting language into the core, I feel that goes against the philosophy of the project. The plugin interface should facilitate this kind of thing well, over RPC.

To get this written down, my original plan was to implement TextMate-compatible syntax highlighting, which is why I did the fancy-regex library (to get Oniguruma-compatible regex parsing). That I was going to put in the core, so that plugins wouldn't be required for basic highlighting. However, I am very open to evidence about other systems which might be better (performance, accuracy, availability of high-quality syntax definitions which can be packaged with appropriate licensing). Thanks for the pointer to the LPeg ones, they very much look worth evaluating.

**Comment by leavengood** [comment]
I understand your feelings on embedding a scripting language in the core. Lua is probably one of the best options though, and LPeg and the infrastructure provided by the [lexer.lua file](https://github.com/martanne/vis/blob/master/lexers/lexer.lua) is pretty nice. I'm not actually a Lua programmer (yet), but this is making me want to learn it a bit.

It is worth trying, but I don't know how well syntax highlighting over RPC would work.

TextMate-compatible is probably good too, as you get to inherit all the existing syntax definitions for TextMate and of course Sublime Text (before they changed the file format.) Sublime's new format is another option I guess.

The main point is to reuse another format so as not to have to reinvent all the syntax definitions, but I'm sure you have already planned for that.

**Comment by leavengood** [comment]
I guess one nice benefit of LPeg is the parsers are composable, so it is easier to support things like CSS or JavaScript inside HTML or HTML and Ruby together in erb files, etc. I don't know how that is handled with the TextMate format.

**Comment by stoivo** [comment]
:+1: for using sublimes syntax files. I have patch some of the new syntax files myself. The new format is a lot easier to work with the the old one. Also wbond is work full time  on sublime, he is fixing and update  sublime syntaxes.

**Comment by leavengood** [comment]
The new Sublime Text format is documented here:

http://www.sublimetext.com/docs/3/syntax.html

**Comment by raphlinus** [comment]
Composable syntax definitions are definitely a feature of TextMate/Sublime. There's a stack, and you can push from one into another, then pop back into the first.

**Comment by rtsuk** [comment]
I'm guessing syntax highlighting will be one of the performance sensitive portions of Xi, so a tightly coupled implemented seems necessary. It will be interesting to see if the regex-based approaches can keep up.

**Comment by leavengood** [comment]
I have started work on a Rust crate to handle TextMate/Sublime style regular expression based syntax highlighting. Actually the crate is more about applying scopes to source code, where syntax highlighting is one particular application of that.

For now I'm starting with how the tmLanguage files work, and later may see if the code can be adapted to the new sublime-syntax files. For now there seems to be more tmLanguage files out in the wild.

I'm using the `onig` crate which wraps the Onigurama regular expression library that these files seem to require.

I think I understand the overall algorithm but some of the details are a bit tricky. For example consider a really stupidly simple set of patterns which match single capital letters and all capital constant names (these are some of the patterns in the [Rust tmLanguage file](https://github.com/rust-lang/sublime-rust/blob/master/Rust.JSON-tmLanguage)):

```json
        {
            "name": "storage.type.source.rust",
            "match": "\\b[A-Z]\\b"
        },
        {
            "name": "constant.other.source.rust",
            "match": "\\b([A-Z_][A-Z0-9_]+)\\b"
        },
```

If you use my current code to parse something like `A CONSTANT 123 B let WORDS` with the two above rules you get scopes for `A`, `B` and `WORDS` but not `CONSTANT`, which should also have one. This is because I stop after the first match is found, and since the `B` matches the first pattern, and I don't keep going to see that `CONSTANT` would be matched by another rule. I then take a substring after the end of the scope found, which is why finding the `B` makes the `CONSTANT` get lost. If I test this same string in Sublime I get all the proper scopes so obviously I'm doing it wrong.

I only have a few hours in this so far so it isn't a big surprise it isn't working right, but my main point is that it is trickier than it looks. Obviously I need to try all the patterns, or try until one is found which matches the beginning of the current string, or find the one which matches closest to the beginning. Overall though it just sounds really inefficient. Also the point of these rules is to order them and the first one which matches is supposed to be the winner, and that consumes that part of the string, so trying all matches seems wrong.

But probably I just need to get something that works and passes a good set of test cases, then I can look into optimizing.

If anyone has suggestions or tips, please let me know.

**Comment by raphlinus** [comment]
@leavengood Some thoughts.

First, I'm pretty sure the matching can be defined by running all the rules, and choosing the one with the earliest match. Anything else leads to chaos, as the rules would be allowed to, eg, match inside comments, etc.

Then, when you get a match, you conceptually run all the rules again, starting at the end position of the last match. This can easily lead to O(n^2) (or worse, depending on exactly how expensive those regexes are to compute). I think there's an optimization that brings this into the realm of reasonably efficient: when you run the regex, you retain the position of first match. Then, when the new start point is before that retained position, you know you don't have to run it again. It's a bunch more state-keeping though.

Finally, you might look into `fancy-regex`, which I created for precisely this purpose. It's not tested yet. If you find any case where it doesn't work well (and it's definitely reasonable to test comparatively against onig, both for correctness and performance), let me know.

Best of luck! I'm hoping to get to the point reasonably soon where I can run syntax highlighting plugins. A lot of the infrastructure is already in place for it, and more is coming (for example, propagating deltas to plugins).

**Comment by leavengood** [comment]
Hey Raph,

As I thought about it my mind definitely started to realize that I need to find the earliest match, even if that means a later rule in the list.

I think I understand your optimization (caching matches later in the string, more or less), but it does sound like quite a bit more state-keeping. I'll see how it goes.

I will also check out fancy-regex.

Thanks for the feedback!

**Comment by drbo** [comment]
I think that any syntax highlighting system needs to be able to handle:

- Non-regular syntax (like Rust itself, which is not even context-free due to its raw strings)
- Nested syntax
- Interpolated variables and expressions in strings
- Syntax error highlighting
- Context-dependent keywords (C#, C++, possibly Rust at some point)
- Externally-provided syntax (Agda, Idris)
- Here-documents, with proper handling of code after the delimiter (sh, Ruby, Perl)

**Comment by chriskrycho** [comment]
Nested syntax is particularly useful when writing *about* software in Markdown: getting syntax highlighting in a code block is very, very helpful.

**Comment by zypeh** [comment]
Is it feasible to do both structured and regex to handle the nested syntax and can be highly extensible also.

Reference: [Wikipedia - Structure editor](https://en.wikipedia.org/wiki/Structure_editor)

**Comment by trishume** [comment]
@raphlinus @leavengood I'm nearly finished a pure Rust library for fast high quality syntax highlighting using Sublime Text themes and syntax definitions: https://github.com/trishume/syntect

Based on discussion in this thread, I think it might work well for Xi. It uses the more modern sublime text syntax file format that @leavengood linked earlier. It currently can parse and highlight using all default Sublime Text syntaxes and themes. There's only some API stability/usability work and some straightforward performance improvements to go. I also designed the system to make super high performance possible, I just have some caching left to implement before it becomes super fast.

I plan on making it fit whatever use case anybody writing a text editor in Rust needs for it to be useful.

Let me know what you think of it as a possible for syntax highlighting in Xi and if there are any specific features or requirements you have for that. I'll start by looking at the fancy-regex library and seeing if it would work better than the `onig` crate.

**Comment by leavengood** [comment]
This looks really awesome @trishume! I started working on my own version but then "real life" intervened, so I'm glad to see someone else took up the torch. I've only briefly looked over your project, but I intend to dig into it more when I get time.

By the way besides being useful as part of an editor, I think a library like this is also useful as a general syntax highlighting tool like [pygments](http://pygments.org), except probably a lot faster (though I would need to verify that.)

Thanks for working on it!

**Comment by trishume** [comment]
@leavengood yah the plan is to add a utility to the API for spitting out HTML for pygments-like use cases. I also can see it being useful for various code intelligence tools, since it spits out quite rich information.

If you benchmark it now it won't be super slow but it won't be fast either. There's three things I have to do and then it should be ultra-fast:
- Cache regex matches so I don't try every regex every time (this is what @raphlinus mentioned and what Textmate does)
- Cache scope selector matches so I don't run full scope selector matching on every token
- Lazily compile syntaxes, currently startup time is bad because it loads and compiles every single syntax file, including thousands of regexes, at startup. Figuring out all the syntaxes is necessary for nesting but I don't have to actually fully compile them and load them until they are needed.

**Comment by raphlinus** [comment]
@trishume From a quick look, that looks great. I was quite ready to implement syntax highlighting myself, but now it certainly looks like I don't have to.

I'm not 100% certain whether syntax highlighting should live in the xi core, or should be a plug-in. I'm leaning toward the latter, which means the focus should be on developing the plug-in interface, specifically getting syntax highlighting off the ground first. I'll definitely play with this more deeply when I have more time (various other things have intervened recently).

I'm very interested to see if fancy-regex works for you, and how it compares with onig.

**Comment by trishume** [comment]
@raphlinus @leavengood I just did some benchmarking and added benchmarks to the readme. It seems currently the highlighting performance (for JS, with syntect using Sublime's ES6 grammar, probably fancier than the other editors) is on par with Textmate 2, Emacs and Visual Studio Code. 3 times faster than Atom and 10 times slower than Sublime Text 3 (and Vim but it uses super basic highlighting). I can probably close some of the gap to Sublime by doing the two caching strats I mentioned, the rest could probably be closed by using fancy-regex to compete with Sublime's custom regex engine.

@raphlinus depending on how fast the RPC and JSON serialization is, I'm pretty sure you'll want it in the core for performance. In order to match Sublime's speed you would need to transmit 10 megabytes/second (including serialization/deserialization) with a custom binary protocol, and maybe 20 megabytes/second with JSON. Twice both those figures if you want parsing to be a separate plugin from highlighting, which you probably would since lots of other things would want parsed scope information. If that's no problem then highlighting in a plugin should be fine.

**Comment by leavengood** [comment]
Cool, that's awesome Tristan. It sounds like `syntect` is already in pretty good shape, even without some of the optimizations you know you can do. It sounds like Sublime Text 3 is pretty awesome though. But given we have @raphlinus and his fancy_regex on our side I'm pretty sure we can get close or even beat Sublime. I think having a strong focus on performance will be a key differentiator for Xi as compared to other open source editors. So I think I can speak for most of us "Xi fans" in saying we appreciate your efforts @trishume to make your syntax highlighting code fast!

As for your second paragraph, given more than just syntax highlighting can use the scoping system (go to symbol, folding, probably more), I think it makes sense to have at least the parsing and scope applying in the Xi core.

It might work to have the actual color applying to scopes for syntax highlighting in a plugin though, but the question would be what benefit does that provide? The only thing I could think is alternative color scheme formats or strategies. But there is still a constraint in that it will still involve applying colors to scopes as opposed to a completely alternative syntax highlighting system like using LPeg. Plus if we support the Sublime or TextMate theme formats we have a much bigger library of color schemes to use, as compared to some new format.

Also now that I've learned the other benefits that the scoping system provides I find it superior to LPeg for syntax highlighting. On that note maybe I should edit the subject of this issue 😄

**Comment by luolong** [comment]
A little tangential perhaps to this discussion but I have always dreamed of an editor where you can easily define your syntax highlighting scheme with pure CSS working on an AST of the document instead of line by line parsing as commonly seen in most of the editors out there. 

With all the freedom that CSS gives in terms of styling. 

For example, one could have slightly different color scheme depending on the nesting level of the code. Or color of the parentheses could reflect the level of nesting. Things like this and probably more that I have not really thought about.

This of course means that breaking up the source into sections with rich metainfo about these sections is really important.

**Comment by luolong** [comment]
Also, maybe I am beating a dead horse here, but what about more of a grammar-like approach to defining the grammars. 

Things like Perl6 grammars would be nice middle ground here but also more formal parser grammars like ANTLR and the like. 

If the api of the highlighter would only pass a source (or a section of a source) to the external plugin which would return with a description of sections in the source, that would allow both styles of syntax highlighting plugins.

**Comment by trishume** [comment]
@luolong 

Sublime text's scope stack and scope selector system is similarly powerful to CSS. See https://manual.macromates.com/en/scope_selectors
Depending on the syntax definition (some are written much better than others) you could do something like highlight different nesting levels different colours.

Sublime's new stack based language which `syntect` implements is actually similarly powerful to full grammars. It's able to parse complex nested structures with no issues. The advantage is that there are already a bunch of existing grammars, and more importantly, it is much easier to write grammars that are robust to incomplete programs (as you often have in an editor) and also much easier to write incomplete grammars. Writing full parsing grammars takes a lot of effort, for some less important languages you might just want to spend a little bit of time to enable more basic syntax highlighting.

On an unrelated note, I've been doing some more optimizing and I now have startup time down to 23ms, which sounds plenty fast to me. It also can load syntax definitions from a compressed binary dump compiled into the library, which means library users don't have to manage their own clone of Sublime's packages.

**Comment by luolong** [comment]
Almost every parser has a parser. Well most of them do.

**Comment by trishume** [comment]
@luolong Since your "Almost every parser has a parser" is tautologically true (except for the almost part) I'm assuming you meant almost every language has a parser.

The thing is every language has a parser written in their own implementation's language that parses into their own AST, which only parses valid code. For highlighting you need a uniform language that outputs a standardized tagging of language elements so that colour schemes can interpret the result and assign the correct colours.

Porting a language parser/grammar to this uniform format is a reasonably big effort, for example the Rust grammar from Sublime Text is 1000 lines long, and `rustc`s `parser.rs` is 5000 lines. All of syntect is only ~2500 lines. All of Sublime's default grammars total 27,000 lines. 

Basically, writing new parsers would be an enormous effort. And if you want to write new parsers or translate existing ones, Sublime's grammar format is a pretty good format to do it to, it maps pretty straightforwardly to a recursive descent parser.

**Comment by luolong** [comment]
Yeah, right. the tautology in my previous comment was an obvious enough typo :) Nicely caught 
(I fixed it... sort of)!

**Comment by leavengood** [comment]
To add another voice to the discussion, @trishume is basically saying that the goals you want for "CSS like styling" @luolong can probably be achieved with the line-by-line scope-based parsing system. You don't need a new parser, you just need a new scope styling system (or maybe the existing one can be made better.) The scoping system is quite capable and is used for many things a language parser is used for, such as "go to symbol" and "code folding".

The current Sublime Text system which `syntect` is implementing is basically a 3rd generation system and has had most of the kinks worked out. It really does this job very well, and starting over with some other kind of parsing system is probably not worth the effort, and furthermore there is no indication it would provide any real benefit.

With all that said, if you (or someone else) wants to try some other approach to parsing languages which can be used for syntax highlighting and has other benefits, I suspect Raph and we other fans of this project would be open to seeing and evaluating such an option.

**Comment by raphlinus** [comment]
The slowest piece of plugin RPC should be JSON deserialization, which the [serde benchmarks](https://github.com/erickt/rust-serialization-benchmarks) suggest is 128MB/s. That seems plenty fast enough compared with the numbers @trishume posted above.

The `syntect` approach looks like it will be a very high bar to exceed, for reason that @trishume has articulated well. That said, I don't think regexes are the One True Way to do syntax highlighting, and I think there's value in keeping the interfaces open-ended so people can experiment with other techniques. One of the things I find disappointing about Sublime Text's plugin API is that it doesn't facilitate doing one's own syntax highlighting.

So what I'd like to do here is build out the plugin API to the point where a plugin can send color spans to the core, and have the first main plugin be syntect (probably with the syntect plugin in the xi repo and automatically installed, so there are no additional building steps). Once we've got it, we can measure the performance, and if it turns out that I'm wrong and RPC is a significant overhead, we can look at streamlining that, possibly by running the plugin as a thread.

The plugin approach requires some rocket science (in particular, getting the async stuff right and tracking the locations of the spans when concurrent edits happen), but that comes from wanting to do syntax highlighting asynchronously, rather than the plugin architecture. It would be possible to get preliminary results earlier by running the highlighting synchronously in the core, but I don't think it's particularly worth doing that. Unfortunately, I've got other commitments for the next couple of weeks that preclude me from having a chunk of time right now, but will be able to dive in after that.

**Comment by luolong** [comment]
An interesting development by Microsoft that might be relevant to this discussion https://github.com/Microsoft/language-server-protocol

**Comment by raphlinus** [comment]
@luolong Thanks for that link, it's very interesting! I tried to find documentation on their protocol a month or so ago, but didn't, probably because I was looking in the Typescript repo.

That document is certainly relevant, but it's also clear to me that it's not well suited to syntax highlighting out of the box. As I design the plugin protocol, I think I will crib liberally from that, but adapt it as needed for real-time highlighting.

**Comment by yaskhan** [comment]
How about that lib: https://github.com/Marwes/combine-language
?

**Comment by raphlinus** [comment]
A heads-up for people following this issue: I've made considerable progress on the plugin protocol recently, to the point where it wouldn't be crazy for someone to start prototyping a plugin for syntax highlighting. I've written a bit more at a [thread on the xi_editor subreddit](https://www.reddit.com/r/xi_editor/comments/4t4vz6/plugin_progress/).

**Comment by raphlinus** [comment]
I just pushed a patch that adds syntect-based highlighting. It's still quite rough, but I'm pleased at the progress. Screenshot below.

<img width="870" alt="xi_syntect_screenshot" src="https://cloud.githubusercontent.com/assets/242367/17075481/8847c8c2-5049-11e6-9709-885c824ce0bf.png">

**Comment by trishume** [comment]
@raphlinus cool, that looks great! Makes a huge difference in how fancy and programmer-editor-like Xi looks, even if there's still a ways left to go.

Just something else that might be of interest if you don't already know: The `Theme` struct has a `settings` attribute that includes all sorts of editor colours like backgrounds, line gutters, highlight colours, etc.. that you can use to make everything fit well together visually. Also see the `find_syntax_by_*` methods on SyntaxSet for syntax detection, but you probably saw those already.

I read the commit, your use of `syntect` looks good, glad you got it working.

**Comment by eyelash** [comment]
@raphlinus I don't know what data structure is used for syntax highlighting but since I read in the commit message that the style spans do not track edits yet I thought this blog post might be helpful http://blog.atom.io/2015/06/16/optimizing-an-important-atom-primitive.html

**Comment by raphlinus** [comment]
@eyelash Good reading, thanks! The data structure is already in place, it's `Spans`, which in turn is an instantiation of the `Tree` that's at the root of `Rope`. It has a lot of similarities to the one described in that post - balanced tree storage, relative indices, etc. And I believe it has the performance characteristics I want.

One small difference from markers: the spans are downstream of undo/redo. So if you edit and then undo, it will recalculate the syntax highlighting. This is because (in the present implementation) undo/redo has some overhead. I might revisit this later, but for syntax highlighting it shouldn't be a problem because it's a pure function of the text.

**Comment by raphlinus** [comment]
@eyelash Commit 1435c59 actually just fixed the tracking. It turned out to be ridiculously simple, not sure why I didn't do it before. That said, wiring up the notifications still remains, as does the OT/CRDT magic for resolving edits that happen concurrently with the plugin calculating spans.

**Comment by rkusa** [comment]
I am actually not sure how serious I am about this, but anyway, I [prototyped](https://github.com/rkusa/xi-golang-plugin) a AST-based syntax highlighting of Go (not all node types have a color yet). The main idea would be to re-use the AST for formatting, listing and syntax highlighting instead of parsing the source for every one of these tasks again.

![screen shot 2016-07-27 at 11 19 58](https://cloud.githubusercontent.com/assets/409021/17170552/e0459fe4-53ec-11e6-8250-9ce4f710a347.png)

This is of course a language-specific thing, but I like the idea of having different syntax highlighting plugins for different source languages.

**Comment by trishume** [comment]
@rkusa Neat, does the Go parser library you're using handle incomplete/invalid code? That's one of the things that bites other attempts to use language parsers in editor plugins. You don't want a parser that gives up after it finds a syntax error, and then everything after that doesn't get highlighted.

**Comment by rkusa** [comment]
@trishume the Go parser is able to handle invalid code to a certain degree. The documentation says:

> The parser accepts a larger language than is syntactically permitted by the Go spec, for simplicity, and for improved robustness in the presence of syntax errors.

However, this will still be a problem, because there are most certainly syntax errors, where the parser will stop. Another challenge is, that the parser is not meant to allow incremental updates and I don't want to end up parsing the whole file after each change. Once we have update notifications to the plugin ready, I'll play around with it a little more; e.g., feeding the parser only with minimal related chunks of code.

**Comment by trishume** [comment]
Cool. The remainder of this comment is actually from last night but I couldn't post it due to the Github outage, hence there is some overlap with what you just mentioned:

Also something potentially relevant to this issue that I realized thanks to @rkusa's prototype: Implementing parsers for other languages in other languages is going to be really difficult to do well.

Not only do you need a parser that can parse quickly and handle invalid/incomplete syntax, in order to be fast you need one that can store intermediate parse states. Then you need to write the logic for incremental re-parsing in your language of choice (unless its Rust, then you can copy the code from the default highlighting plugin when that gets done). If you don't do that your parser is going to be slower than the default parser for large files.

I don't actually think this is a big deal though since I'm a fan of the power of sublime-syntax files and they highlight almost every language quite well. There's also either sublime-syntax or tmLanguage (Sublime has a tmLanguage->sublime-syntax converter) files for almost every language people use.

**Comment by rkusa** [comment]
I definitely agree that there is the need for a syntax highlighter like syntect. The point that I wanted to support, is the beauty of providing syntax highlighting through plugins instead of baking it into the core. I am also expecting that the sublime syntax approach will end up being faster compared to the AST approach for Go, but I'll continue experimenting with it anyway, 'cause it's possible ;-)

**Comment by raphlinus** [comment]
Sounds like we're all in agreement here - the syntect-based plugin will serve the needs of most people, and can be made to work well in the near term. But I still really like the idea of having it in a plugin, to allow (nay, encourage) experimentation with other methods.

I plan to close this issue fairly soon (basically, as soon as #110 lands and gets some cleanup). There is still work to be done: performance improvement, support for bg color spans, etc., but those can be tracked separately.

**Comment by jansol** [comment]
This can probably be closed now?
